File: 2020/day2/part2/main.py
from typing import List

from advent_of_code.day2.part1.main import make_policy, Policy
from advent_of_code.helpers.helpers import read_file_line_by_line


def is_password_valid(p: Policy) -> bool:

    i: List[int] = [p.lower_limit - 1, p.upper_limit - 1]  # indexes

    if (
        (p.password[i[0]] != p.character and p.password[i[1]] == p.character)
        or
        (p.password[i[0]] == p.character and p.password[i[1]] != p.character)
    ):
        return True
    return False


def main(file: str = "") -> int:
    f: List[str] = read_file_line_by_line(filepath=file)

    results: List[bool] = []

    for line in f:
        policy = make_policy(line=line)
        results.append(is_password_valid(p=policy))

    count: int = 0 if not results.count(True) else results.count(True)

    return count

File: 2020/day2/part2/main.go
package part2

import (
	"advent-of-code/advent_of_code/day2/part1"
	"advent-of-code/advent_of_code/helpers"
	"fmt"
)

func IsPolicyValid(p part1.Policy) bool {
	i := []int{p.LowerLimit - 1, p.UpperLimit - 1} // indexes
	char1 := string(p.Password[i[0]])
	char2 := string(p.Password[i[1]])

	if char1 != p.Character && char2 == p.Character {
		return true
	}
	if char1 == p.Character && char2 != p.Character {
		return true
	}
	return false
}

func Main(filepath string) int {
	lines, err := helpers.ReadFileLineByLine(filepath)
	if err != nil {
		fmt.Println(err)
	}

	var results int

	for _, line := range lines {
		policy := part1.MakePolicy(line)
		if IsPolicyValid(*policy) {
			results++
		}
	}

	return results

}
